A5_LogOn.ino

```diff
@@ -0,0 +1,303 @@
+  char CharacterReceived;
+  uint8_t Length;
+
+  //if there is another character
+  if(Serial.available() > 0)
+  { 
+    //get character
+    CharacterReceived = Serial.read();
+
+    //cases per character
+    switch(CharacterReceived)
+    {
+      case '\n':
+        //terminate password string
+        password[LogOnIndex] = '\0';
+
+        //add end of table record
+        LogOnRecord[LogOnIndex+1] = 0x91;
+        
+        //save password record in EEPROM
+        WriteEeprom(LogOnAddress, LogOnRecord, LogOnIndex+2);
+
+        //print password
+        PrintPassword();
+        
+        //next
+        LogOnAddress += LogOnIndex+1;
+        StartTableAddress = LogOnAddress;
+        LogOnState = 7;
+      break;
+      
+      //password character
+      default:
+        //if buffer is not full
+        Length = LOGON_LENGTH;
+        if(LogOnIndex < Length)
+        {
+          //save character in password string
+          password[LogOnIndex] = CharacterReceived;
+          
+          //save character in password record
+          LogOnRecord[LogOnIndex+1] = CharacterReceived;
+
+          //next
+          LogOnIndex += 1;
+        }
+        else
+        {
+          //dump any other characters typed
+          while(Serial.available() > 0)
+          { 
+            //get character
+            Serial.read();
+          }
+          
+          //print error message
+          LogOnLengthError();
+
+          //try again
+          LogOnState = 3;
+        }
+      break;
+    }
+  }
+}
+
+void GetSSID(void)
+{ 
+  char Character;
+  uint8_t Length;
+
+  //if buffer is not empty
+  Length = LOGON_LENGTH;
+  if(LogOnIndex < Length)
+  {  
+    //get next character
+    Character = LogOnRecord[LogOnIndex+1];
+
+    //cases per character
+    switch(Character)
+    {
+      //password record ID
+      case 0x83:
+        //terminate ssid string
+        ssid[LogOnIndex] = '\0';
+        
+        //print SSID
+        PrintSSID();
+        
+        //next
+        LogOnAddress += LogOnIndex+1;
+        ReadLogOnRecord();
+        LogOnIndex = 0;
+        LogOnState = 6;
+      break;
+      
+      //ssid character
+      default:
+        //save character in ssid string
+        ssid[LogOnIndex] = Character;
+        
+        //next
+        LogOnIndex += 1;
+      break;
+    }
+  }
+  else
+  {
+    //ask user to enter
+    LogOnState = 1;
+  }
+}
+
+void GetPassword(void)
+{
+  char Character;
+  uint8_t Length;
+
+  //if buffer is not empty
+  Length = LOGON_LENGTH;
+  if(LogOnIndex < Length)
+  {  
+    //get next character
+    Character = LogOnRecord[LogOnIndex+1];
+
+    //cases per character
+    switch(Character)
+    {
+      //start of pattern record ID
+      case 0x90:
+        //terminate password string
+        password[LogOnIndex] = '\0';
+        
+        //print Password
+        PrintPassword();
+        
+        //next
+        LogOnAddress += LogOnIndex+1;
+        StartTableAddress = LogOnAddress;
```

Comment:
I understand the desire to economize space and pack the pattern records as close to the logon records as possible, but I think that ultimately the better is to allocate a constant amount of space for the logon records, and always start the pattern records at the same address. This means that there will usually be some unused space between the end of the logon records and the start of the pattern records, but this prevents the user from having to reupload the pattern information every time the he inputs a new ssid/password.